Lots of rewriting for the Scala 3 Reference (#15993)
lampepfl/dotty: docs/_docs/reference/changed-features/structural-types.md

## Example

> Before we can use structural types, we need a subclass of `Selectable` that overrides `selectDynamic` if we want `a.b` and `applyDynamic` if we want `a.b(c)` (more details further down). As an example, here is a class `Record`, that only provides `selectDynamic`:

Reviewer: I think this language isn't very precise. For  _selection_ from a structural type, the type needs to subtype `Selectable`. However structural refinements can be to any type

> Before we can use structural types, we need a subclass of `Selectable` that overrides: `selectDynamic` if we want `a.b` and `applyDynamic` if we want `a.b(c)` (more details further down). As an example, here is a class `Record`, that only provides `selectDynamic`:

Reviewer 2: ```suggestion
Before we can use structural types, we need a subclass of `Selectable` that overrides `selectDynamic` if we want `a.b` and `applyDynamic` if we want `a.b(c)` (more details further down). As an example, here is a class `Record`, that only provides `selectDynamic`:
```

```scala
class Record(elems: (String, Any)*) extends Selectable:
  private val fields = elems.toMap
  def selectDynamic(name: String): Any = fields(name)
 ```

Now that we have `Record`, we can define a structural type `Person`:

```scala
type Person = Record { val name: String; val age: Int }
 ```

The type `Person` adds a _structural refinement_ to its parent type `Record` that defines the two fields `name` and `age`. We say the refinement is _structural_ since  `name` and `age` are not defined in the parent type. But they exist nevertheless as members of class `Person`. 

To construct instances of this type, we have to go through `Record`, and use an explicit cast:

```scala
val person = Record("name" -> "Emma", "age" -> 42).asInstanceOf[Person]
```

Once this is done, `person` behaves as we expect from `Person`: they have two fields `name` and `age` that can be accessed normally.
For instance, the following program would print  "Emma is 42 years old.":

```scala
println(s"${person.name} is ${person.age} years old.")
```

> As `Person` doesn't have a member `address`, the following returns a compile-time error (regardless of how `person` was constructed):

Author: ```suggestion
As `Person` doesn't have a member `email`, the following returns a compile-time error (regardless of how `person` was constructed):
```

  private val fields = elems.toMap
  def selectDynamic(name: String): Any = fields(name)
```
Which allows us to create instances of `Person` like so:
```scala
val person = Record("name" -> "Emma", "age" -> 42).asInstanceOf[Person]
```

The parent type `Record` in this example is a generic class that can represent arbitrary records in its `elems` argument. This argument is a
sequence of pairs of labels of type `String` and values of type `Any`.
When we create a `Person` as a `Record` we have to assert with a typecast
that the record defines the right fields of the right types. `Record`
itself is too weakly typed so the compiler cannot know this without
help from the user. In practice, the connection between a structural type
and its underlying generic representation would most likely be done by
a database layer, and therefore would not be a concern of the end user.

Besides `selectDynamic`, a `Selectable` class sometimes also defines a method `applyDynamic`. This can then be used to translate function calls of structural members. So, if `a` is an instance of `Selectable`, a structural call like `a.f(b, c)` would translate to

```scala

>   a.applyDynamic("f")(b, c)

Reviewer: ```suggestion
a.applyDynamic("f")(b, c)
```

```

## Using Java Reflection

Using `Selectable` and [Java reflection](https://www.oracle.com/technical-resources/articles/java/javareflection.html), we can select a member from unrelated classes.

> Before resorting to structural calls with Java reflection one should consider alternatives. For instance, sometimes a more a modular _and_ efficient architecture can be obtained using [type classes](https://docs.scala-lang.org/scala3/book/types-type-classes.html).

> For example, we would like to provide behavior for both [`FileInputStream`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/FileInputStream.html#%3Cinit%3E(java.io.File)) and [`Channel`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/channels/Channel.html) classes by caling their `close` method, however, these classes are unrelated, i.e. have no common supertype with a `close` method. Therefore, below we define a structural type `Closeable` that defines a `close` method.

Reviewer: ```suggestion
For example, we would like to provide behavior for both [`FileInputStream`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/FileInputStream.html#%3Cinit%3E(java.io.File)) and [`Channel`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/channels/Channel.html) classes by calling their `close` method, however, these classes are unrelated, i.e. have no common supertype with a `close` method. Therefore, below we define a structural type `Closeable` that defines a `close` method.
```

the Scala compiler to:

```scala
person.selectDynamic("name").asInstanceOf[String]
person.selectDynamic("age").asInstanceOf[Int]
```

Besides `selectDynamic`, a `Selectable` class sometimes also defines a method `applyDynamic`. This can then be used to translate function calls of structural members. So, if `a` is an instance of `Selectable`, a structural call like `a.f(b, c)` would translate to

```scala
a.applyDynamic("f")(b, c)
```

## Using Java Reflection

> If a structural type's base type does not have Selectable as a supertype, as below, it uses [Java reflection](https://www.oracle.com/technical-resources/articles/java/javareflection.html) instead:

Author: ```suggestion
If a structural type's base type does not have `Selectable` as a supertype, it uses [Java reflection](https://www.oracle.com/technical-resources/articles/java/javareflection.html) instead.
If no base type is present, the base type is `Any`. 
**Note:** `Any` is not a subtype of `Selectable`, so base-type-less structural types will always use Java reflection.
Example:
```

Author: @sjrd / @bishabosha Is the above clear ?

Reviewer: I think this implies that Java reflection is automatic - this is not the case, I think the original unchanged sentence is most clear

> Write about {} meaning with an Any at the beginning

Reviewer 2: ?

Author: Sorry that was a TODO that slipped through
```suggestion

```

  try op(f) finally f.close()
```

The call `f.close()` requires `Closeable` to extend `Selectable` to identify and call the `close` method in the receiver `f`. A universal implicit conversion to `Selectable` is enabled by an import
of `reflectiveSelectable` shown above, based on [Java reflection](https://www.oracle.com/technical-resources/articles/java/javareflection.html). What happens "under the hood" is then the following:

>  - The implicit wraps `f` in an instance of `scala.reflect.Selectable` (which is a subtype of `Selectable`).

Reviewer: ```suggestion
 - The implicit conversion wraps `f` in an instance of `scala.reflect.Selectable` (which is a subtype of `Selectable`).
```

>      reflectiveSelectable(f).applyDynamic("close")()

Reviewer: it might be strange that this is the only place where we suddenly introduce `applyDynamic`, previously it looked like there was only `selectDynamic`

Author: I think it's acceptable, as the example requires it, and `applyDynamic` is introduced as early as line 84
